modules/reactor/src/meshgenerators/PinMeshGenerator.C

```diff
@@ -478,9 +480,58 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
   else
     declareMeshProperty("extruded", false);
 
+  if (getReactorParam<bool>("generate_rgmb_metadata"))
+    generateMetadata();
+
   _build_mesh = &getMeshByName(build_mesh_name);
 }
 
+void
+PinMeshGenerator::generateMetadata()
+{
+  std::string mg_struct = _is_assembly ? "assembly" : "pin";
+  std::string metadata_prefix = mg_struct + "_" + std::to_string(_pin_type);
```

Comment:
As I am attempting to gather this metadata by following the naming conventions here, I think it would make more sense for the prefix to start with the name of the mesh generator (applicable to the assembly and core mesh generators too). Presumably those are all unique names.

Ie, this pin generator block below would prompt all metadata to start with `fuel_pin_a` instead of `pin_1`.
```
  [fuel_pin_a]
    type = PinMeshGenerator
    reactor_params = rmp
    pin_type = 1
    pitch = ${fuel_pin_pitch}
    num_sectors = 2
    ring_radii = '${fuel_clad_r_i} ${fuel_clad_r_o}'
    mesh_intervals = '1 1 1'    # Fuel, cladding, background
    region_ids = '${mid_lower_refl}     ${mid_lower_refl}     ${mid_lower_refl};
                  ${mid_fuel_1}         ${mid_ht9}            ${mid_sodium};
                  ${mid_upper_na_plen}  ${mid_upper_na_plen}  ${mid_upper_na_plen};
                  ${mid_upper_gas_plen} ${mid_upper_gas_plen} ${mid_upper_gas_plen}' # Fuel, cladding, background
    quad_center_elements = false
  []
```
That would make it easier for me to gather the metadata in an action since I readily know the mesh generator names but need to know the metadata string explicitly to be able to call the value.

Reply:
In a similar vein, I would also suggest removing "pin_type" from the prefix and setting that as it's own attribute: 
```
metadata_prefix = [name of mesh generator block]
declareMeshProperty(metadata_prefix + "_pin_type", std::to_string(_pin_type));
```

Reply:
I'm personally not a huge fan of such a design, as it would complicate the definition of the assembly / pin lattice as we'd also have to store the names of each of the constituent pins / assemblies to indicate where they correspond to in the lattice as well. I'm in favor of metadata having a structure that is more hierarchal (it is clear which values correspond to pins, assemblies, and core) and agnostic to the way the user defines each of the pins, assemblies, lattice etc. in the input file. 

I also don't think naming metadata the way you are suggesting will solve your problem of knowing what the mesh generator name is supposed to be, since retrieving metadata requires you to still provide the name of the mesh generator on which the metadata is stored (i.e. the action isn't able to query metadata defined on the mesh without also providing a mesh generator name).

@GiudGiud thoughts on this? @kkiesling is trying to define an action that queries this metadata outside of the mesh generation stage to define the MC CSG geometry, and so doesn't have a way to get the final mesh generator name that has all the metadata stored onto it, unless this information is passed directly to the action itself.

Reply:
To clarify - I can get the mesh generator names just fine, but I cannot get the value of any metadata without knowing the exact string it goes by (using `getMeshProperty()`).

Reply:
Given the current metadata naming structure, what names wouldn’t be able to be able to be inferred if you had a mesh generator name?

Reply:
It makes it an unnecessary layer to go through. Right now I need to read all this info to be able to construct the prefix from the mesh generator params, not the mesh directly. If I have to read half of the parameters already, why not construct a CSG model from the input file directly rather than from a mesh? But perhaps I am going about trying to gather this metadata in a completely wrong way.

Reply:
Not really sure I understand what you mean by "read all this info", but If you can pass the metadata prefix of the starting mesh you are trying to construct to the action as well, i.e. "core", "assembly_<assembly_type_id>", or "pin_<pin_type_id>", then all the metadata that you need to construct the CSG mesh can be inferred.

For example, if you are starting with a core, there will be a metadata name "core_assembly_types" which gives you a list of assembly type ids that constitute the core, and you know that `assembly_<assembly_type_id>_<assembly_metadata_name>` will also be defined on the mesh. Similarly there is a value "core_pin_types" that will give you the pin type ids that are defined within the core (although you would probably infer the pin types from the assemblies themselves).

If your starting mesh that you are trying to generate is an assembly with type id <assembly_type_id>, then there will be a metadata name `assembly_<assembly_type_id>_pin_types` that will give you the pin types defined in the assembly and you know that `pin_<pin_type_id>_<pin_metadata_name>` will also be defined on the mesh.

In case this is not clear, all of this metadata will be saved on a single mesh generator name, so as long as you have the name of the starting mesh you are trying to create, you won't have to worry about any other mesh generator names to obtain all metadata information. Let me know if there is additional information that I am missing that you would need.

Reply:
I think we should have a quick chat about this. I think the new metadata capability for propagation of props will help us here. And in this case - we want things to be hierarchical.

Reply:
@shikhar413 [here](https://github.com/kkiesling/moose/blob/929fa5d0deb3522e6e903e4ba8be7c3161a0b177/modules/reactor/src/actions/MonteCarloGeomAction.C#L93C38-L93C38) is an example of what needs to be read from the actual input for the meshgenerator if using this prefix setup just to be able to get the full metadata name to be able to call it (and that is after already having to know the meshgenerator name). `pin_type` and `use_as_assembly` in this context are required to be called using `getParam` and then the full prefix can be known before calling any mesh properies. 

I was thinking about this more over the weekend and I am honestly wondering if it is necessary to have any prefix on the metadata at all? (Disclaimer that this is my first experience with moose metadata so I could be completely coming at this from the wrong angle or be totally unaware of already existing norms on this). But if all the metadata has to be called from the meshgenerator which has a unique name, I wonder why we have to use the prefix to specify that? My mental model of the metadata is that it is like an attribute on Python object. All objects of the same type have attributes callable by the same variable name and the value is what differs, but you know the uniqueness of that attribute because it came from the unique object. So in this case, each meshgenerator is like the unique object and why not just call each piece of metadata by the same variable name across all so they are easily accessible? In this pin case, every `PinMeshGenerator` would have `pin_type`, `use_as_assembly`, `pitch`, `radii`, etc as the full name of the metadata.

Reply:
So I think things _could_ be done this way, I'm not sure if this is the "MOOSE" way to do things this way, and frankly I'm not sure if I'd be able to comment on what makes the most sense here, since the design of the metadata storage system will be undergoing some updates, as @loganharbour alluded to. Generally, however, mesh generators work in a modular way where they are only supposed to act on the information available directly in the "inputs" parameter (i.e. core meshes only retrieve information that is available on the input assemblies, assembly meshes only retrieve information that is available on the input pins), and I've been propagating relevant metadata up to the necessary level if say a core needed pin-level information. Mesh generators do not store the "history" of how they got to the current point in the mesh generation workflow, they take an input mesh and modify it in some way so that they can be passed into a successive mesh generator.

Of course the metadata system works slightly different than how mesh generators work, since you're able to query metadata on any mesh generator instead of relying on information that is directly available through the inputs parameter, but my only concern with relying on using mesh generator names instead of prefixes to define metadata is it now assumes that there is some mesh generator in the history with a pre-defined name, and relying on such an assumption could be more error-prone and would require more checks in the metadata store (for context, I have an application that uses this metadata and I'd need to be storing more information w.r.t the mesh generator names to make sure that the input mesh does indeed have the correct "history" of mesh generators, instead of checking the metadata defined on a single input core mesh generator).

I'll let @GiudGiud and @loganharbour chime in here, since they are more knowledgeable about the design of the Mesh Generator and metadata system as a whole

Reply:
I would appreciate a call to hash this out and understand better how the metadata system is being updated. I don't really see how knowing the meshgenerator's name is an issue. It is already required in the current setup when trying to gather the metadata.

Reply:
>  It is already required in the current setup when trying to gather the metadata.

It won't be in the future. MGs are supposed to be very plug and play, with most generators working with other generators without needing to know information about the upstream generators. Of course - this isn't the case with reactor, but we still want our very general generators (renames, transforms, etc) to work with this module when possible. With this, you add in the ability to put non-reactor MGs in-between reactor MGs. Which.. no longer allows you to rely on the input MG name for upstream metadata, because that input MG could be a non-reactor MG.

Reply:
Marking this as resolved per our discussion, let me know if there are changes to the code / metadata naming that you would like to see now